Dev

vnebelunka opened the pull request.

vnebelunka committed:
initial commit

vnebelunka committed:
vectors for fields added

vnebelunka commented on main_field_ideal.cpp:
```diff
@@ -0,0 +1,112 @@
+
+int main(int argc, char **argv)
+{
+    const std::string input_dir = argc > 1 ? argv[1] : "./data/", \
```
лучше работать с аргументами командной строки, чем в коде прописывать адреса.
При смене геометрии не надо пересобирать

vnebelunka commented on main_field_ideal.cpp:
```diff
@@ -0,0 +1,112 @@
+    //Открытие файла с точками
+    std::ifstream fin(result_dir + "body.gr");
+    if (!fin.is_open()) {
+        throw std::runtime_error("Can't opent file: " + result_dir + "body.gr");
```
Лучше пробрасывать exception, чтобы при падении программы выскакивало что-то более явное, чем тождественная единица

vnebelunka commented on main_field_ideal.cpp:
```diff
@@ -0,0 +1,112 @@
+//======================Parameters=============================
+    ED_Par ed_param(input_dir + "ed_param.txt");
+    //E0
+    E0_Polar_Type e0_H(0, ed_param.k_vec);
```
параметры с конечным числом значений (тип поляризации) лучше делать через enum

AlenaSetukha replied:
enum class

vnebelunka commented on include/Num_Par.h:
```diff
@@ -9,6 +9,6 @@ class Num_Par
     double kappa, M;
 
-    Num_Par(const char* filename);
+    Num_Par(const std::string &filename);
```
везде файлы передавались как std::string, а тут случился char* почему-то

AlenaSetukha replied:
+

vnebelunka commented on include/get_field.h:
```diff
@@ -29,14 +30,17 @@
 
-void get_field_ideal(const double*** cells, const double** norm,
-        const std::complex<double>** j_vec,
+template<size_t CellPoints>
```
универсальная функция по количеству точек в геометрии (3 или 4)

AlenaSetukha replied:
+

vnebelunka commented on include/get_field.h:
```diff
@@ -29,14 +30,17 @@
+template<size_t CellPoints>
+void get_field_ideal(const std::vector<double[CellPoints][3]> &cells, const std::vector<double[3]> &norm,
+        const std::vector<std::complex<double>[3]> &j_vec,
         const double max_diag, const int num_frm,
```
`const int num_frm` теперь лишний аргумент, векторвы сами содержат свой размер

AlenaSetukha replied:
+

vnebelunka commented on include/get_field.h:
```diff
@@ -118,9 +121,9 @@ void save_field_as_gv(const std::string filename_out,
     {
         fout_u_real << field[i][0].real() << " " << field[i][1].real() << " " << field[i][2].real() << std::endl;
         fout_u_im << field[i][0].imag() << " " << field[i][1].imag() << " " << field[i][2].imag() << std::endl;
-        u_abs = sqrt(abs_tmp(field[i][0]) * abs_tmp(field[i][0]) +
```
непонятно, чем обычный `std::abs` не устроил.
Я бы в других файлах тоже выпилил эту функцию

AlenaSetukha replied:
+

vnebelunka commented on include/constants.h:
константы точности, такие как `machine_zero`, лучше делать через std::numeric_limits.
Также namespace лучше, чем класс (у тебя всё равно этот класс всегда одинаково заполняется).
В таком случае константы лучше делать constexpr, а не конст

vnebelunka commented on include/get_einc.h:
отдельный файл на одну функцию плодит очень много файлов.
У меня всякие такие вещи собранны в одном ED-utilities

vnebelunka commented on include/ED_Par.h:
1. class public == struct
2. я бы сделал конструктор не только из файла, но просто по пачке аргументов (как это сделано с параметрами интегрирования, например).
3. А ещё отделил бы параметры, связанные с ЭПР (постобработкой задачи) и свойствами исходной задачи

vnebelunka commented on src/K.cpp:
часть параметров передано как complex*, а часть как complex[3], хотя все они размера 3
* `const std::complex<double> (&j)[3]`
* `std::complex<double>* res`

AlenaSetukha replied:
теперь все &[3]

vnebelunka commented on src/ED_Par.cpp:
return в конце конструкторов и void функций не нужен

AlenaSetukha replied:
Убрала

vnebelunka commented on src/element_geom.cpp:
1. `vec_prod` -> template.
2. res у void функций, то указателем, то ссылкой (в контексте базовых переменных лучше ссылкой всегда)
3. `div_mod` - ??? по коду это не деление, а что-то вроде `return a * (a != b)`

AlenaSetukha replied:
1. В функциях типа vec_prod два аргумента могут быть в разном порядке: (double, double), (complex, double), (double, complex), (complex, complex). 
Если сделать два темплейта, не понятно, каким определить результат. Это старый вопрос, решили оставить четыре функции.
2. Про res не поняла. Если возвр. значение процедуры одно, то &res, если значений много - то массив *res.
3. div_mod - ненужная функция(аналог деления по модулю, но только для a <= b).Ее убрала.

vnebelunka replied:
1. `
template<typename T1, typename T2>
void vec_prod(const T1 *a, const T2 *b, decltype((*a) * (*b)) *res);
`
2. пон

vnebelunka commented on src/get_einc.cpp:
из-за того, что constants у тебя не namespace, а class, тебе приходится в функция плодить абсолютно одинаковые объекты.

AlenaSetukha replied:
+

vnebelunka committed:
minor changes for compilation

vnebelunka committed:
merged with main

vnebelunka commented on include/get_j.h:
```diff
@@ -14,12 +15,12 @@
 
 //-------------------------Дополнительная функция считывания готовых токов-----------------------
-void get_j_from_files(const std::string &filename_real, const std::string &filename_image, std::vector<std::complex<double>[3]> &j_vec);
+void get_j_from_files(const std::string &filename_real, const std::string &filename_image,
```
Строки, как и любые большие объекты, лучше передавать по ссылке

vnebelunka commented on src/common_type.cpp:
```diff
@@ -15,92 +16,89 @@
+    std::ifstream fin2(filename);
+    if (!fin2.is_open()) {
         std::cout << "Read geodat.dat error" << std::endl;
-        exit(-1);
```
exception

vnebelunka commented on src/element_geom.cpp:
```diff
@@ -149,16 +125,8 @@ double tr_square(const double* pnt_1, const double* pnt_2, const double* pnt_3)
     return res;
 }
 
-//======================================Vector length============================================
```
общую реализацию template нельзя обьявлять в компилируемом файле, т.к само описание template не генирирует никакого кода (пока не произойдёт инстанцирование).
Перенёс реализацию в header

AlenaSetukha replied:
везд вставить реализацию темплэйта в хэдер

vnebelunka commented on src/get_einc.cpp:
```diff
@@ -5,14 +5,14 @@
 {
-    Constants c;
-    std::complex<double> deg = pow(c.e, c.i_complex * scal_prod(k_vec, x));
+    std::complex<double> deg = pow(Constants::e, Constants::i_complex * scal_prod(k_vec, x));
```
exp(...), complex<double>(0, 1)

vnebelunka committed:
constexpr constants

vnebelunka committed:
minor optimizations

vnebelunka commented on include/integral_universal_pnt.h:
```diff
@@ -33,13 +33,11 @@ void integral_universal_pnt(const double (&x)[3], const double (&rut0)[4][3],
-    P* ff = new P[int_param.idim];
-    P* res_prev = new P[int_param.idim];
+    P* ff = new P[int_param.idim]();
+    P* res_prev = new P[int_param.idim]();
```
вызывает дефолтный конструктор для new

vnebelunka commented on src/kernel_lib.cpp:
```diff
@@ -94,9 +95,9 @@ void f_double_pot_L(const double (&x)[3], const double (&y)[3], const f_par& par
 
 
         ff[0] = Constants::pi_reverse * scal_prod(norm, diff) / r / r / r;//K(x,y)
-        if (r >= Constants::machine_zero && r < param.rs)
```
фиктивное условие с учетом предыдущего ифа

vnebelunka committed:
Helmholtz simple potential optimization

vnebelunka committed:
numerical stability minor change

vnebelunka committed:
bugfix

AlenaSetukha committed:
K_rot_rot и R_rot
Написаны функции операторов K_rot_rot и R_rot в различных случаях
для расчета поля от электрических токов
	- вне тела в дальней зоне
	- вне тела в дальней зоне с уточнением
	- в точках коллокаций
Операторы:
	- K_rot_rot_Far - дальняя зона, K = пов. инетграл от специального ядра
	- K_rot_rot - ближняя зона без выделения особенности
	- K_rot_rot_Near - ближняя зона с выделением особенности
	- R_rot_Far - дальняя зона, обычное сглаживание
	- R_rot_Colloc - оператор R точках коллокации(учет = 0 при i=j)
Немного переделаны классы:
	integral_par, f_par, ED_par
Добавлены аналитические интегралы.

vnebelunka commented:
@AlenaSetukha не коммить `build` файлы, пожалуйста
Они платформозависимые, а при чтении изменений надо пролистывать все эти 100 файлов с огромными числом изменений.
На гите обычно лежат только исходники, но в нашем случае ещё входные данные (на самом деле их лучше выносить в отдельный репозиторий, но лень)
Инструмент, чтобы не следить каждый раз за лишними машинными файлами: [.gitignore](https://www.atlassian.com/git/tutorials/saving-changes/gitignore)

vnebelunka committed:
generates removed

vnebelunka committed:
.gitignore added

vnebelunka committed:
global constants fix

vnebelunka committed:
constant power of 10 removed
